ls1intum/Artemis
# `General`: Change password handling for internal users

### Warning: Data Migration
All passwords of internal users will get hashed. Reverting the version will not be possible without resetting the database.
Before starting the application with the new version
- Create a database backup
- Configure `artemis.bcrypt-salt-rounds` in your Spring configuration with the help of https://github.com/ls1intum/bcrypt-Benchmark to use the most secure setting




### Checklist
#### General


- [X] I tested **all** changes and their related features with **all** corresponding user types **locally** as TS testing is not possible here.
#### Server
#### Client
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types **locally** with the Atlassian Suite.
- [x] I tested **all** changes and their related features with **all** corresponding user types **locally** with Jenkins and Gitlab.

### Motivation and Context


This is the final PR of my Password Migration to Bcrypt Hashing to finally switch the Password Service.

### Description

I deprecated the old PasswordService and added a new one supporting Bcrypt. A new Java Migration sets all passwords of external users to null and first decrypts and then hashes the other passwords to complete the transformation.
I then adapted all occurrences, especially all instances of creating and updating the user. Instead of decrypting the password at any given place I now pass the password along.
I also documented the new usage and current limitations of Jenkins.

I upped the maximum missable classes because by deprecating the password service I created another class that is, again, just a wrapper class for external functionality so nothing we need or want to write tests for.

**The cypress tests currently fail. I can't reproduce certain errors locally. Currently investigating if this PR requires the cypress setup to do some changes.**

### Steps for Testing

You can only test this locally as this change would make any test server unusable until reset or this PR is merged.
For local testing:
1. SQL dump `jhi_user`
2. Do the testing
3. Drop `jhi_user` and import your SQL dump
4. Remove the entry from the `migragion_changelog` table

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Start the application
2. Make sure that `jhi_users` was updated correctly
3. Log in with internal and external users
5. Do a password change with internal users (verify the password gets updated for the git repository)
6. Edit internal users with a password change and without. Verify that changes are correctly propagated. For Jenkins developers, please validate that the behaviour was correctly documented.
7. Test anything that seems reasonable to you in regards to the changes.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





SecurityConfiguration.java: n/a | 97.65%
JiraAuthorizationInterceptor.java: 100% | 100%
ArtemisInternalAuthenticationProvider.java: 41.67% | 74.07 %
DomainUserDetailsService.java: 0% | 36.36%
BitbucketUserManagementService.java: 71.05% | 94.34%
VcsUserManagementService.java: n/a | 100%
GitLabUserManagementService.java: 88.24% | 97.50%
JenkinsUserManagementService.java: 86.84% | 98.64%
PasswordService.java: n/a | 100%
UserCreationService.java: 68.97% | 94.12%
UserService.java: 78.33% | 93.25%
AccountResource.java: 77.08% | 95.71%
UserResource.java: 75% | 96.77%
